aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,886 @@
+
+        for (int jj = Ap[i];  jj < Ap[i + 1]; jj++) {
+          indices_row[jj] = i;
+        }
+      });
+
+  // Filling the COO column indices 
+  thrust::for_each(
+    policy,
+    thrust::make_counting_iterator(int64_t(0)),
+    thrust::make_counting_iterator(int64_t(csr_output.nnz_)),
+    [output_indices_accessor,
+      csr_output_pointers_accessor,
+      csr_output_ind_accessor,
+      major_dim,
+      nnz] __device__(int64_t i) {
+      int64_t* indices_col = output_indices_accessor[1].data();
+      indices_col[i] = csr_output_ind_accessor[i];
+    });
+}
+
+} // end anonymous namespace
+
+Tensor sparse_matrix_mask_helper_cuda(
+  int64_t r_nnz,
+  const SparseTensor& t,
+  const LongTensor& mask_indices
+) {
+  auto t_v = t._values().contiguous();
+
+  Tensor r_values = at::empty({r_nnz}, t_v.options());
+
+  auto t_i = t._indices().contiguous();
+  auto t_indices_accessor = t_i.packed_accessor<int64_t, 2>();
+  auto t_nnz = t._nnz();
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  Tensor t_flatten_indices = at::empty({t_nnz}, mask_indices.options());
+  auto t_flatten_indices_accessor = t_flatten_indices.packed_accessor<int64_t, 1>();
+  auto t_n_cols = t.size(1);
+
+  thrust::for_each(
+    policy,
+    thrust::make_counting_iterator(int64_t(0)),
+    thrust::make_counting_iterator(int64_t(t_nnz)),
+    [t_indices_accessor, t_flatten_indices_accessor, t_n_cols] __device__ (int64_t i) mutable {
+      auto index = t_indices_accessor[0][i] * t_n_cols + t_indices_accessor[1][i];
+      t_flatten_indices_accessor[i] = index; 
+  });
+
+  Tensor mask_flatten_indices = at::empty({r_nnz}, mask_indices.options()); 
+  auto mask_flatten_indices_accessor = mask_flatten_indices.packed_accessor<int64_t, 1>();
+  auto mask_indices_accessor = mask_indices.packed_accessor<int64_t, 2>();
+
+  thrust::for_each(
+    policy,
+    thrust::make_counting_iterator(int64_t(0)),
+    thrust::make_counting_iterator(int64_t(r_nnz)),
+    [mask_flatten_indices_accessor, mask_indices_accessor, t_n_cols] __device__ (int64_t i) mutable {
+      auto index = mask_indices_accessor[0][i] * t_n_cols + mask_indices_accessor[1][i];
+      mask_flatten_indices_accessor[i] = index; 
+  });
+  auto max_sz = std::max(r_nnz, t_nnz);
+  Tensor t_index_set = at::empty({max_sz}, mask_indices.options()); 
+
+  auto result_end = thrust::set_intersection_by_key(
+    policy,
+    t_flatten_indices.data_ptr<int64_t>(),
+    t_flatten_indices.data_ptr<int64_t>() + t_nnz, 
+    mask_flatten_indices.data_ptr<int64_t>(), 
+    mask_flatten_indices.data_ptr<int64_t>() + r_nnz,
+    thrust::make_counting_iterator(int64_t(0)),
+    thrust::make_discard_iterator(),
+    t_index_set.data_ptr<int64_t>());
+
+  auto new_sz = thrust::distance(t_index_set.data_ptr<int64_t>(), result_end.second);
+  
+  if (new_sz > 0) {
```

Comment:
This won't work if the computed gradient does not have all the indices that the mask has. In this case, given that we require that gradient sparsity pattern is the same as original tensor, we still need `r_values` set to zero there, so `r_values` has to be initialized as `zeros` and not `empty`.

Reply:
Thanks for the feedback, I was about to solve this CUDA issue.